torch/utils/data/_utils/pin_memory.py

```diff
@@ -53,27 +53,13 @@ def do_one_step():
         do_one_step()
 
 def pin_memory(data, device=None):
-    if isinstance(data, torch.Tensor):
-        return data.pin_memory(device)
-    elif isinstance(data, (str, bytes)):
-        return data
-    elif isinstance(data, collections.abc.Mapping):
```

Comment:
Only types that are explicitly registered will be considered as pytree nodes.

```python
class MyTuple(tuple):
    pass

tree = MyTuple((tensor1, tensor2))  # is a leaf node

# before
pin_memory(tree)  # pin for tensor1 and tensor2

# use pytree
pytree.tree_map_(pin_memory, tree)  # no effect
```

I think the correct refactoring is:

```diff
+   elif isinstance(data, collections.abc.MutableMapping):
+       try:
+           clone = data.copy()
+           clone.update({k: pin_memory(sample, device) for k, sample in data.items()})  # type: ignore[call-arg]
+           return clone
+       except TypeError:
+           # The mapping type may not support `__init__(iterable)`.
+           return {k: pin_memory(sample, device) for k, sample in data.items()}
    elif isinstance(data, collections.abc.Mapping):
        try:
            return type(data)({k: pin_memory(sample, device) for k, sample in data.items()})  # type: ignore[call-arg]
        except TypeError:
            # The mapping type may not support `__init__(iterable)`.
            return {k: pin_memory(sample, device) for k, sample in data.items()}
```

Reply:
Gotcha
OOC why doesn't pytree consider subtypes? That seems intuitive to me and relatively safe no?

Reply:
> Gotcha OOC why doesn't pytree consider subtypes? That seems intuitive to me and relatively safe no?

1. Performance: use `type(obj) is registered_type` or `type(obj) in TYPE_REGISTRY` is much faster than:

```python
def is_leaf(obj):
    for cls in TYPE_REGISTRY:
        if isinstance(obj, cls):
            return False
    return True
```

2. Signature of the constructor: we need to reconstruct a pytree in `tree_unflatten` function, which needs to call the constructor of the node type. But not all subtypes provide the same interface of `__new__` or `__init__`. For example, `tuple` and `NamedTuple` types:

```python
seq: Iterable[Any]

new_tuple = tuple(seq)
new_namedtuple = NamedTupleClass(*seq)
```

3. What if someone registers `class object` as node type? All types will be nodes.

Reply:
For the first point why not just `isinstance(obj, REGISTERED_TYPES)`? Why do we need to iterate over the classes?

For the second point I agree but in a way it won't work now if you have not registered namedtuple and it won't work after you use `isinstance`, so it's not breaking things right?

For the third I guess the lowest class in the register should prevail, ie namedtuple > tuple > object

Reply:
My concern is that in python generally if I subclass something I expect that this will only come with extra features, not fewer. And pytree is a rather private feature of PyTorch, many people use/will use it without knowing it. Having strict type assessments can create surprising behaviours if someone just builds a thin wrapper around tuple, dict or list.

Reply:
> For the first point why not just `isinstance(obj, REGISTERED_TYPES)`? Why do we need to iterate over the classes?

It's still an $O(n)$ operation while `type(obj) in REGISTERED_TYPES` is $O(1)$. Note that `REGISTERED_TYPES` is a dictionary.

> For the second point I agree but in a way it won't work now if you have not registered namedtuple and it won't work after you use `isinstance`, so it's not breaking things right?

Here is another example:

```python
class Foo:
    def __init__(self, x=None):
        self.x = x

register_node(
    Foo,
    lambda foo: ([foo.x], None),
    lambda values, context: Foo(values[0]),
)

class Bar(Foo):
    def __init__(self):  # different signiture of the paraent class
        super().__init__(x=None)
        del self.x

bar1 = Bar(x=1)  # error
bar2 = Bar()
isisntance(bar2, Foo)  # -> True
bar2.x  # error

tree_map(lambda x: x, Bar())  # type changed to Foo after unflattening back
```

> For the third I guess the lowest class in the register should prevail, ie namedtuple > tuple > object

So we need to implement a hierarchy for registered types. It is too complex and introduces new performance issues. Also, it will be buggy for multi-inherit cases.